File: archive/ruby/muse/data_structures/ArrayListTest.rb
require "muse/data_structures/ArrayList"
require "muse/util/TestRunner"

def testArrayList
  size = 8192

  list = ArrayList::ArrayList.new(0)

  1.upto(size) do |i|
    list.append(i)
  end

  list.shrink

  return false if list.size != size

  return false if list.capacity != size

  0.upto(size - 1) do |i|
    return false if list.get(i) != i + 1
  end

  0.upto(size - 1) do |i|
    list.set(i, size - i)
  end

  0.upto(size - 1) do |i|
    return false if list.get(i) != size - i
  end

  i = 0
  j = size - 1
  while i < j
    x = list.get(i)
    y = list.get(j)

    list.remove(i)
    list.insert(i, y)
    list.remove(j)
    list.insert(j, x)
    i += 1
    j -= 1
  end

  0.upto(size - 1) do |i|
    return false if list.get(i) != i + 1
  end

  size.downto(1) do |i|
    return false if list.back != i

    list.eject
  end

  list.shrink

  return false unless list.isEmpty

  list.capacity.zero?
end

TestRunner.pick(testArrayList) if __FILE__ == $PROGRAM_NAME

File: archive/php/muse/data_structures/ArrayListTest.php
<?php

declare(strict_types=1);

require_once "muse/data_structures/ArrayList.php";
require_once "muse/util/TestRunner.php";

use muse\data_structures\ArrayList;
use muse\util\TestRunner;

function testArrayList(): bool
{
    $size = 8192;

    $list = new ArrayList(0);

    for ($i = 1; $i <= $size; $i++) {
        $list->append($i);
    }

    $list->shrink();

    if ($list->size() != $size) {
        return false;
    }

    if ($list->capacity() != $size) {
        return false;
    }

    for ($i = 0; $i < $size; $i++) {
        if ($list->get($i) != $i + 1) {
            return false;
        }
    }

    for ($i = 0; $i < $size; $i++) {
        $list->set($i, $size - $i);
    }

    for ($i = 0; $i < $size; $i++) {
        if ($list->get($i) != $size - $i) {
            return false;
        }
    }

    for ($i = 0, $j = $size - 1; $i < $j; $i++, $j--) {
        $x = $list->get($i);
        $y = $list->get($j);

        $list->remove($i);
        $list->insert($i, $y);
        $list->remove($j);
        $list->insert($j, $x);
    }

    for ($i = 0; $i < $size; $i++) {
        if ($list->get($i) != $i + 1) {
            return false;
        }
    }

    for ($i = $size; $i >= 1; $i--) {
        if ($list->back() != $i) {
            return false;
        }
        $list->eject();
    }

    $list->shrink();

    if (!$list->isEmpty()) {
        return false;
    }

    return $list->capacity() == 0;
}

if (count(debug_backtrace()) == 0) {
    TestRunner\pick("testArrayList");
}
